Malte lig 1262 advanced selection external docs

MalteEbner opened the pull request:
## Description

- adds docs for selection using the new config
- adapts existing docs (getting started, docker active learning)
- already included: DIVERSIFY->DIVERSITY

### Tabs vs. Dropdowns
Here tabs are used for the SelectionStrategy and Dropdowns for the configuration examples.
![image](https://user-images.githubusercontent.com/20324507/186690356-6aae8b68-707b-4639-af80-d625959a60af.jpeg)

philippmwirth commented:
As an initial test I went through the "First Steps" section and I have some feedback. Will summarize it below.

**Feedback:**
- "Scheduling a Simple Job"
    - I copy pasted the command and got a syntax error (there's a missing comma)
    - After fixing it, it complained that all the imports are missing (-> add imports)
    - Then I had the problem that the imports didn't work because I had the wrong PIP version (-> clarify)
    - Then I got `TypeError: __init__() got an unexpected keyword argument 'type'` because there's a typo:
    ```python
           SelectionConfigEntry(
                input=SelectionConfigEntry(type=SelectionInputType.EMBEDDINGS),
                strategy=SelectionConfigEntry(type=SelectionStrategyType.DIVERSIFY)
            )
    ```
    should be
    ```python
           SelectionConfigEntry(
                input=SelectionConfigEntryInput(type=SelectionInputType.EMBEDDINGS),
                strategy=SelectionConfigEntryStrategy(type=SelectionStrategyType.DIVERSIFY)
            )
    ```
    - Finally, I managed to run it (it's running in the background now).
 - The typos mentioned above can also be found in "Training a Self-supervised Model"
 - In the example for "Training a Self-supervised Model" with exposed lightly arguments there's no selection config
 - The typos mentioned above can also be found in "Specifying Relevant Files"
- In the "Selection":
    - I don't think this is the proper way to use the word "tell":
    > The input tells on which data the objective is defined on. It defines a scalar number or vector for each image in the dataset.
    - I don't think we support custom embeddings so why is it documented so prominently?
    - The formatting in the "Selection Input" and "Selection Strategy" chapters is strange, just leave the `input=` away.
    - For the "Configuration Examples": Don't use tabs - it looks really bad.
    - I don't think the object frequency scorer works the way it's documented. As you can see [here](https://github.com/lightly-ai/lightly/blob/3fbc70267e6993ab8f8556746e2f4314fe5cf3df/lightly/active_learning/scorers/detection.py#L13) it's always normalized to `[0, 1]` so thresholding `> 2` doesn't make sense.


@IgorSusmelj please note that some of this feedback might be outdated after the discussion we had with @MalteEbner @guarin and @michal-lightly about how to pass the selection config (dict vs object).


I will provide more feedback when testing a more involved example but in general we need to make sure that all the examples and code snippets work!

IgorSusmelj commented:
Suggestion for blocks you can test:

- [ ] First steps (/docker/getting_started/first_steps.html)
- [ ] Selection until configuration examples (docker/getting_started/selection.html)
- [ ] Selection from (including) configuration examples to end (docker/getting_started/selection.html)
- [ ] Advanced
  - [ ] Datapool + pretagging (/docker/advanced/datapool.html + /docker/advanced/pretagging.html)
  - [ ] Active Learning + Object Level (/docker/advanced/active_learning.html + /docker/advanced/object_level.html) 
  - [ ] Sequence selection (/docker/advanced/sequence_selection.html)
- [ ] Integration (/docker/integration/dagster_aws.html)
- [ ] Examples (/docker/examples/overview.html)

philippmwirth reviewed (commented):
Reviewed "First steps (/docker/getting_started/first_steps.html)" so far. Good news is that I only had to fix one thing and everything ran through 🙂 

One thing I noticed: The percentage stopping condition is not documented anywhere.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+    .. tab:: EMBEDDINGS
+
+        If you don't provide your own
+        embeddings, the `lightly OSS framework for self supervised learning <https://github.com/lightly-ai/lightly>`_ is used. It generates 32-dimensional
+        embeddings with default settings using self-supervised learning. 
```
I would remove the part about "32-dimensional embeddings" and "default settings". I would only write "lightly framework for self supervised learning is used to generate embeddings.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+        Alternatively, you can use **lightly_pretagging** as the task to use object detections created by the Lightly Worker itself.
+        See :ref:`docker-pretagging` for reference.
+
+        The supported score types are explained here :ref:`lightly-active-learning-scorers`.
```
"here" -> "here:"

or "here" -> "in"

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+
+        .. _worker-selection-predictions:
+
+        The class distribution probability vector of predictions can be used as well. Here, three case have to be distinguished:
```
case -> cases

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+
+            - **Object Detection** → The probability vector of the class predictions of all objects in an image are summed up.
+
+            - **Object Detection** and using the **Object Level** workflow → Each sample is a cropped object and has a single object prediction, whose probability vector is used.
```
link to object level workflow would be great

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+            Use as key the “path” you specified when creating the metadata in the datasource.
+
+
+            **Lightly Metadata**, on the contrary, is calculated out of image data on the fly. It is specified by appending a :code:`lightly` to the key.
```
**Lightly Metadata** is automatically calculated by the Lightly Worker.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+            "strategy": {
+                "type": "BALANCE",
+                "target": {
+                    "Ambulance": 0.2, # `Ambulance` should be a valid class in your `schema.json`
+                    "Bus": 0.4
+                }
```
Same here, make it add up to 100%. We can always explain later on what happens if it does not add up to 100%.

MalteEbner commented:
> One thing I noticed: The percentage stopping condition is not documented anywhere.

Because it is not implemented yet: https://linear.app/lightly/issue/LIG-1661/add-proportion-samples-to-openapi-specs

guarin reviewed (commented):
Partial review for the `getting_started/selection.rst` part. Will finish it tomorrow :)

I tried the code and it worked (with some minor tweaks). I'll also try all the "configuration examples" later on.

Edit: It looks like @philippmwirth already tested the remaining part so I'll skip it.

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+                        },
+                        "strategy": {
+                            "type": "THRESHOLD",
+                            "threshold": 2,
```
the score is normalized to `[0, 1]` so this doesn't work the way it's intended to

IgorSusmelj replied:
Damn. I thought I removed this example

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+
+        .. code-block:: python
+
+            {
```
here it's important to switch `pretagging=True`

philippmwirth reviewed (commented):
Reviewed the "Getting Started -> Selection" part. All the tests run through when switching `n_samples` to `nSamples` but I have some more feedback.

guarin commented on docs/source/docker/advanced/code_examples/python_run_datapool_example.py:
```diff
@@ -2,6 +2,8 @@
 import lightly
 from lightly.openapi_generated.swagger_client.models.dataset_type import DatasetType
 from lightly.openapi_generated.swagger_client.models.datasource_purpose import DatasourcePurpose
+from lightly.openapi_generated.swagger_client import SelectionInputType, SelectionStrategyType, SelectionConfig, \
+    SelectionConfigEntry, SelectionConfigEntryInput, SelectionConfigEntryStrategy
```
These imports are unused:
```
from lightly.openapi_generated.swagger_client import SelectionInputType, SelectionStrategyType, SelectionConfig, \
    SelectionConfigEntry, SelectionConfigEntryInput, SelectionConfigEntryStrategy

```

guarin commented on docs/source/docker/advanced/pretagging.rst:
```diff
@@ -86,13 +86,13 @@ Lightly Worker config: `pretagging=True`
 
 - `pretagging=True` enables the use of the pretagging model
 - `pretagging_debug=True` add a few images to the report for debugging showing the image with the bounding box predictions.
```
I just saw that further down in the document there are more config options that do not have a `:code:`

guarin commented on docs/source/docker/advanced/code_examples/python_run_active_learning.py:
```diff
@@ -75,15 +78,48 @@
         "enable_training": False,
         "pretagging": False,
         "pretagging_debug": False,
-        "method": "coral",  # we use the coral method here
-        "stopping_condition": {
```
Remove also the following comments further up:
```
# Schedule the docker run with 
#  - "active_learning.task_name" set to your task name
#  - "method" set to "coral"
```

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+- (Optional)  :ref:`Model predictions <docker-datasource-predictions>` such as classifications, object detections or segmentations
+- (Optional) :ref:`Custom metadata <docker-datasource-metadata>` can be anything you can encode in a json file (from numbers to categorical strings)
+
+.. warning:: Using the selection config is a new feature and **selecting a 
```
In the docker: yes
In the API: no, see https://linear.app/lightly/issue/LIG-1661/add-proportion-samples-to-openapi-specs

MalteEbner replied:
I added it, a the issue was merged.

guarin reviewed (commented):
Next review :)

guarin commented on docs/source/docker/examples/datasets_in_the_wild.rst:
```diff
@@ -112,11 +112,19 @@ selects a diverse set of frames:
+                  "strategy": {
+                      "type": "DIVERSIFY"
+                  }
+              }
+          ]
```
The sentence on line 132 needs an update as well: "To perform a random selection we can simply replace "coreset" with "random" as our selection method. Note that Coreset is the default method."

guarin commented on docs/source/docker/examples/academic_datasets.rst:
```diff
@@ -51,18 +51,29 @@ selects a diverse set of frames:
 
 .. code-block:: python
 
+    # in this example we use a diversifying selection strategy (CORESET)
```
Line 48: frames -> images

guarin commented on docs/source/docker/examples/academic_datasets.rst:
```diff
@@ -51,18 +51,29 @@ selects a diverse set of frames:
+                    },
+                    "strategy": {
+                        "type": "DIVERSIFY",
+                        "stopping_condition_minimum_distance": 0.1
```
The old command did not use a minimum distance.

guarin reviewed (commented):
Final review 😅

japrescott commented on docs/source/docker/examples/academic_datasets.rst:
```diff
@@ -51,18 +51,28 @@ selects a diverse set of frames:
+            "pretagging_debug": False
+        },
+        selection_config = {
+            "nSamples": 500000,
```
feels weird that its `nSamples` and not stopping condition as its with min_distance.

MalteEbner replied:
We cannot really have a "stopping condition" struct anymore, as the minimum distance stopping condition was moved to the DIVERSIFY strategy, as it only applies to that strategy and is not global anymore.

japrescott commented on docs/source/docker/getting_started/first_steps.rst:
```diff
@@ -248,16 +255,6 @@ After the job was processed, the selected data will be accessible in the configu
 
-There's an alternative stopping condition to `n_samples`, the `min_distance`
-
-- :code:`stopping_condition.min_distance` **0.2** would remove all samples which are
```
is this no longer true? Maybe move it to where we use it in one of the examples

MalteEbner replied:
We now have an example for it when explaining the DIVERSIFY strategy:

```json
"strategy": {
    "type": "DIVERSIFY",
    "stopping_condition_minimum_distance": 0.2
}
```

japrescott replied:
I was more referring to the removal of the information as it was not moved to a new location

MalteEbner replied:
In my eyes, the concept of the `stopping_condition_minimum_distance` is explained well enough in the diversity section. Thus I see this as solved.
How would you  explain/show it @japrescott

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+        ]
+    }
+
+The variable :code:`n_samples` must be a positive integer specifying the absolute number of samples which should be selected.
```
again n_samples

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+
+Here are examples for the full configuration including the input for several objectives:
+
+.. tabs::
```
See the updated PR description after changing to dropdown

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+
+    .. tab:: EMBEDDINGS
+
+        If you don't provide your own
```
See https://github.com/lightly-ai/lightly/pull/900#discussion_r949172187

MalteEbner replied:
I updated the description accordingly.

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+        ]
+    }
+
+The variable :code:`n_samples` must be a positive integer specifying the absolute number of samples which should be selected.
```
Now n_samples

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+
+Each of these objectives is defined by a `strategy`. A strategy consists of two parts:
+
+- The input tells on which data the objective is defined on. It defines a scalar number or vector for each image in the dataset.
```
I changed it to

> The :code:`input` defines which data the objective is defined on. This data is either a scalar number or a vector for each sample in the dataset.

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,533 @@
+            Use as key the “path” you specified when creating the metadata in the datasource.
+
+
+            **Lightly Metadata**, on the contrary, is calculated out of image data on the fly. It is specified by appending a :code:`lightly` to the key.
```
Yes, see https://linear.app/lightly/issue/LIG-1734/advanced-selection-test-all-available-lightly-metadata

MalteEbner replied:
I added the three types currently supported.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+
+        The class distribution probability vector of predictions can be used as well. Here, three cases have to be distinguished:
+
+            - **Image Classification** → The probability vector of each sample's prediction is used directly.
+
+            - **Object Detection** → The probability vectors of the class predictions of all objects in an image are summed up.
+
+            - **Object Detection** and using the :ref:`docker-object-level` → Each sample is a cropped object and has a single object prediction, whose probability vector is used.
```
It is a bit special that we use arrows ("→") here as we don't use them anywhere else in the docs

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+        
+        If you use your own predictions (see :ref:`docker-datasource-predictions`), the task name and class names are taken from the specification in the prediction `schema.json`.
+        
+        Alternatively, you can use the Lightly pretagging and the class names are specified here: :ref:`docker-pretagging`. In that case, the task name is `lightly_pretagging`.
```
Alternatively, you can use the Lightly pretagging and the class names are specified here: :ref:`docker-pretagging`.

This sentence is a bit weird, I don't think it is grammatically correct.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+                }
+
+            Currently supported metadata are :code:`sharpness`, :code:`snr` (signal-to-noise-ratio) and :code:`sizeInBytes`.
+            If you use case would profit from more metadata computed out of the image, please reach out to us.
```
you --> your

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+        embeddings in the output dataset. Since we normalize the input embeddings
+        to unit length, this value should be between 0 and 2.0.
+        This is often a convenient method when working with different data sources and trying to combine them in a balanced way.
+        If you want to use this stopping condition to stop the selection early, make sure that you allow selecting enough samples by setting :code:`n_samples` high enough.
```
setting :code:`n_samples` high enough. --> setting :code:`n_samples` or :code:`proportion_samples` high enough.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+    .. tab:: WEIGHTS
+
+        The objective of this strategy is to **select samples that have a high numerical value**. 
+        It requires the input to be a Nx1 matrix of numbers.
```
I would drop the "Nx1" sentence again.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+    .. tab:: THRESHOLD
+
+        The objective of this strategy is to only **select samples that have a numerical value fulfilling a threshold criterion**.
+        E.g. they should be bigger or smaller than a certain value. Like for weighting, this strategy requires the input to be a Nx1 matrix of numbers.
+
```
Drop Nx1 sentence

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+
+.. dropdown:: Metadata Thresholding
+
+    This can be used e.g. to remove blurry images, which equals choosing
+    samples whose sharpness is over a threshold:
+
```
e.g. to --> to e.g.

guarin replied:
choosing --> selecting

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+            ]
+        }
+
+    .. note:: To use the `lightly pretagging` you need to enable it using :code:`'pretagging': True` in the
+              worker config.
```
Would be great to add a link to the lightly pretagging page as well.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+-------------------------
+
+Generally, the order in which the different strategies were defined in the config does not matter.
+In a first steps, all the thresholding strategies are applied.
```
steps --> step

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+In the next step, all other strategies are applied in parallel.
+
+.. note:: Note that different taskes can also be combined. E.g. you can use predictions 
+          from "YOUR_IMAGE_CLASSIFICATION_TASK" for one strategy combined with predictions from
+          "YOUR_OBJECT_DETECTION_TASK" from another strategy.
```
Until now we have always written the task names in lower case.

guarin commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+- **Tradeoff between different objectives.**
+  The optimizer always has to tradeoff between different objectives.
+  E.g. it may happen that all samples with high WEIGHTS are close together. If you also specified the objective DIVERSITY, then only a few of these high-weight samples
+  may be chosen. Instead, also other sample that are visually more diverse, but have lower weights, are chosen.
```
sample --> samples

guarin reviewed (commented):
Almost there! There are a few typos and minor errors left but otherwise this looks great!

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+            "input": {
+                "type": "PREDICTIONS",
+                "task": "my_object_detection_task",
+                "name": "CLASS_DISTRIBUTION"
```
Currently, you can only set it to `CLASS_DISTRIBUTION`.
In the future, it will allow to use other parts of the predictions. E.g. we could have "CLASS_DISTRIBUTION_OBJECTNESS_WEIGHTED" which would weight the sum of class distributions of each prediction by the objectless score of this probability. It is just preparation for the future.

MalteEbner commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,521 @@
+
+        Setting :code:`"stopping_condition_minimum_distance": 0.2` will remove all samples which are
+        closer to each other than 0.2. 
+        This allows you to specify the minimum allowed distance between two image
+        embeddings in the output dataset. Since we normalize the input embeddings
+        to unit length, this value should be between 0 and 2.0.
```
No, we don't.

MalteEbner replied:
I updated the description to make this more clear.

MalteEbner commented:
I would remove the whole "active_learning.rst" part as
- we do not have a specific active learning workflow anymore, it is part of the general selection
- it is still in the docker archive for those that need it

MalteEbner commented on docs/source/docker/advanced/active_learning.rst:
```diff
@@ -33,20 +33,18 @@ In order to do active learning with Lightly, you will need the following things:
+In this example we use a combination of three selection strategies.
 
-Here's an example of how to configure an active learning run:
+- **DIVERSITY** the selected samples using embeddings
```
See https://github.com/lightly-ai/lightly/pull/900#issuecomment-1236721676

MalteEbner commented on docs/source/docker/advanced/active_learning.rst:
```diff
@@ -33,20 +33,18 @@ In order to do active learning with Lightly, you will need the following things:
 
-Here's an example of how to configure an active learning run:
+- **DIVERSITY** the selected samples using embeddings
+- **WEIGHTS** samples with a larger uncertainty using predictions
+- **THRESHOLD** samples based on metadata
```
See https://github.com/lightly-ai/lightly/pull/900#issuecomment-1236721676

MalteEbner commented on docs/source/docker/examples/academic_datasets.rst:
```diff
@@ -170,11 +191,20 @@ the new images.
-            }
+        },
+        selection_config = {
+            "n_samples": 10_000, # maket this number high (upper limit)
```
I changed it to `"proportion_samples": 1.0,`

philippmwirth commented:
> I removed the whole "active_learning.rst" part as
> 
>     * we do not have a specific active learning workflow anymore, it is part of the general selection
> 
>     * it is still in the docker archive for those that need it

This will result in a dead link which is not optimal. Can you add a redirect to the selection page at least?

MalteEbner commented:
> > I removed the whole "active_learning.rst" part as
> > ```
> > * we do not have a specific active learning workflow anymore, it is part of the general selection
> > 
> > * it is still in the docker archive for those that need it
> > ```
> 
> This will result in a dead link which is not optimal. Can you add a redirect to the selection page at least?

I added a redirect:
- using the sphinx-reredirects package: https://documatt.gitlab.io/sphinx-reredirects/usage.html
- tested it: `localhost:1234/docker/advanced/active_learning.html` gets redirected to `localhost:1234/docker_archive/advanced/active_learning.html`

From the logs:
```
::1 - - [05/Sep/2022 15:52:27] "GET /docker/advanced/active_learning.html HTTP/1.1" 304 -
::1 - - [05/Sep/2022 15:52:27] "GET /docker_archive/advanced/active_learning.html HTTP/1.1" 304 -
```

Please be aware that docker/advanced will NOT include active learning anymore, as this will be part of the regular selection.

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,518 @@
+
+Lightly allows you to specify the subset to be selected based on several objectives.
+
+E.g. you can specify that the images in the subset should be visually diverse, be images the model struggles with (Active Learning),
```
Either don't use capital letters for active learning or make it a link

MalteEbner replied:
I switched to lowercase.

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,518 @@
+
+Lightly allows you to specify several objectives at the same time. The algorithms try to fulfil all objectives simultaneously.
+
+Lightly's data selection algorithms are supporting four types of input:
```
Lightly's data selection algorithms support four types of...

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,518 @@
+
+Lightly's data selection algorithms are supporting four types of input:
+
+- **Embeddings** computed using `our open source framework for self-supervised learning <https://github.com/lightly-ai/lightly>`_
```
It's visually unappealing that some of the points are bold and others are links

MalteEbner replied:
Igor drafted this and I think that it is OK.

MalteEbner replied:
Alternatively, we could also delete it, as it contains the same information as later in the detailed input description.
Perhaps @IgorSusmelj has an opinion here.

philippmwirth replied:
your call 🙂

MalteEbner replied:
I would keep it like this and change it when we see the need for it.

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,518 @@
+Prerequisites
+-------------
+
+For using the selection, you must have
```
the selection? I'd rephrase: "In order to use the selection feature, ..."

also must is a bit of a strong word

MalteEbner replied:
I changed it to `In order to use the selection feature, you need to`

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,518 @@
+----------------------------------------------
+
+For scheduling a Lightly Worker run with a specific selection,
+you can use the python client and its :py:`schedule_compute_worker_run` method.
```
this link doesn't render correctly

MalteEbner replied:
![image](https://user-images.githubusercontent.com/20324507/188821934-2d51e7bd-35d1-46d5-a804-bfce276f73d6.jpeg)

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,518 @@
+    }
+
+The variable :code:`n_samples` must be a positive integer specifying the absolute number of samples which should be selected.
+Alternative to :code:`n_samples`, you can also set :code:`proportion_samples` to set the number of samples to be selected relative to the input dataset size.
```
Alternatively

philippmwirth commented on docs/source/docker/getting_started/selection.rst:
```diff
@@ -0,0 +1,518 @@
+
+        Can be used with **SCORES** and **numerical METADATA**. It is specified as follows:
+
+        .. code-block:: python
```
for me it's still not clear what's kept and what's removed here

MalteEbner replied:
I agree, it is explained better in the usage example. I added an extra sentence:
`This will keep all samples whose value (specified by the input) is >= 20 and remove all others.`

philippmwirth reviewed (commented):
I have a bit more feedback.

I also have a question: Do you know if @guarin tested out all the examples?

MalteEbner commented on docs/source/docker/advanced/code_examples/python_run_active_learning.py:
```diff
@@ -75,15 +73,48 @@
+            },
+            {
+                "input": {
+                    "type": "PREDICTIONS",
```
I deleted that script, as it was only reference by `active_learning.rst`
It is still included in the `docker_archive`.

MalteEbner commented:
> I also have a question: Do you know if @guarin tested out all the examples?

I assume that he has, as there are ticks for all the tests where he put his name.

philippmwirth commented on docs/source/docker/advanced/code_examples/python_run_object_level.py:
```diff
@@ -73,15 +73,33 @@
+                }
+            },
+            {
+                "input": {
```
hm ok for me this is a bit confusing and I'd prefer to have only object level in the object level example

MalteEbner replied:
What do you think about adding a comment: "Optionally, you can combine this with active learning to prefer selecting objects the model struggles with."

philippmwirth replied:
would be a good start :) if this is the outcome of some discussion I wasn't part of I also don't want to block this any further

MalteEbner replied:
I added that:

```python
  # Optionally, you can combine diversity selection with active learning
  # to prefer selecting objects the model struggles with.
  # If you want that, just include the following code:
  """
  {
      "input": {
          "type": "SCORES",
          "task": "vehicles_object_detections", # change to your task
          "score": "uncertainty_entropy" # change to your preferred score
      },
      "strategy": {
          "type": "WEIGHTS"
      }
  }
  """
  ```

michal-lightly reviewed (approved):
Just approving since other people are OOO.
